backend/alerts/integrations/microsoft-teams/microsoft_teams.go

```diff
@@ -0,0 +1,234 @@
+		return nil, nil, errors.New("REACT_APP_FRONTEND_URI not set")
+	}
+
+	var oauthEndPoint = oauth2.Endpoint{
+		AuthURL:   fmt.Sprintf("%s/oauth2/v2.0/authorize", authBaseUrl),
+		TokenURL:  fmt.Sprintf("%s/%s/oauth2/v2.0/token", authBaseUrl, tenantID),
+		AuthStyle: oauth2.AuthStyleInParams,
+	}
+
+	options := []oauth2.AuthCodeOption{
+		oauth2.SetAuthURLParam("grant_type", "client_credentials"),
+		oauth2.SetAuthURLParam("scope", "https://graph.microsoft.com/.default"),
+	}
+
+	return &oauth2.Config{
+		ClientID:     clientID,
+		ClientSecret: clientSecret,
+		Endpoint:     oauthEndPoint,
+		RedirectURL:  fmt.Sprintf("%s/callback/microsoft_teams", frontendUri),
+	}, options, nil
+}
+
+func GetAccessToken(ctx context.Context, tenantID string) (*oauth2.Token, error) {
+	conf, opts, err := GetOAuthConfigForTenant(tenantID)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return conf.Exchange(ctx, "", opts...)
+}
+
+func GetAadGroupIDFromActivity(activity schema.Activity) string {
+	team, ok := activity.ChannelData["team"]
+	if ok {
+		team, ok := team.(map[string]interface{})
+		if ok {
+			teamID, ok := team["aadGroupId"].(string)
+			if ok {
+				return teamID
+			}
+		}
+	}
+	return ""
+}
+
+func doGetRequest[T any](accessToken string, url string) (T, error) {
+	return doRequest[T]("GET", accessToken, url, "")
+}
+
+func doRequest[T any](method string, accessToken string, url string, body string) (T, error) {
+	var unmarshalled T
+	client := &http.Client{}
+
+	// code to tell whether we are using absoluteUrl or relative url
+	var finalUrl = fmt.Sprintf("%s%s", MicrosoftGraphUrl, url)
+	parsedUrl, err := nUrl.Parse(url)
+
+	if err != nil {
+		return unmarshalled, errors.Wrap(err, "error parsing url") // this really should not happen
+	}
+
+	if parsedUrl.IsAbs() {
+		finalUrl = url
+	}
+
+	req, err := http.NewRequest(method, finalUrl, strings.NewReader(body))
+	if err != nil {
+		return unmarshalled, errors.Wrap(err, "error creating api request to microsoft")
+	}
+
+	if accessToken != "" {
+		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", accessToken))
+	}
+	if method != "GET" {
+		req.Header.Set("Content-Type", "application/json")
+	}
+
+	res, err := client.Do(req)
+	if err != nil {
+		return unmarshalled, errors.Wrap(err, "error getting response from microsoft endpoint")
+	}
+
+	b, err := io.ReadAll(res.Body)
+	if res.StatusCode != 200 && res.StatusCode != 201 {
+		return unmarshalled, errors.New("Microsoft Graph API responded with error; status_code=" + res.Status + "; body=" + string(b))
+	}
+
+	if err != nil {
+		return unmarshalled, errors.Wrap(err, "error reading response body from microsoft endpoint")
+	}
+
+	err = json.Unmarshal(b, &unmarshalled)
+	if err != nil {
+		return unmarshalled, errors.Wrap(err, "error unmarshaling microsoft api response"+string(b))
+	}
+
+	return unmarshalled, nil
+}
+
+func GetMicrosoftTeamsChannelSuggestions(workspace *model.Workspace) ([]*model.MicrosoftTeamsChannel, error) {
```

Comment:
This feels overkill and I also return the first error we encounter.
Below is a simpler non-concurrent version I wrote

```go
func GetTeamsChannel(workspace *model.Workspace) ([]*model.MicrosoftTeamsChannel, error) {
	allChannels := []*model.MicrosoftTeamsChannel{}
	teamsGroups := GetMicrosoftTeamsGroupsFromWorkspace(workspace)

	for _, teamGroup := range teamsGroups {
		channels, err := GetMicrosoftTeamsChannels(*workspace.MicrosoftTeamsTenantId, teamGroup)
		if err != nil {
			return nil, err
		}
		allChannels = append(allChannels, channels...)
	}

	return allChannels, nil
}
```
I feel like there is a simpler way to do this but I'm unable to find one at the moment.
You can rewrite this or suggest a "better" way of going about this.